lib/services/watcher_test.go

```diff
@@ -0,0 +1,203 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package services_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/jonboulle/clockwork"
+	"github.com/stretchr/testify/require"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/backend/lite"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/services/local"
+)
+
+func TestProxyWatcher(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+
+	bk, err := lite.NewWithConfig(ctx, lite.Config{
+		Path:             t.TempDir(),
+		PollStreamPeriod: 200 * time.Millisecond,
+	})
+	require.NoError(t, err)
+
+	type client struct {
+		services.Presence
+		types.Events
+	}
+
+	presence := local.NewPresenceService(bk)
+	w, err := services.NewProxyWatcher(ctx, services.ProxyWatcherConfig{
+		ResourceWatcherConfig: services.ResourceWatcherConfig{
+			Component:   "test",
+			RetryPeriod: 200 * time.Millisecond,
+			Client: &client{
+				Presence: presence,
+				Events:   local.NewEventsService(bk, nil),
+			},
+		},
+		ProxiesC: make(chan []types.Server, 10),
+	})
+	require.NoError(t, err)
+	t.Cleanup(w.Close)
+
+	// Since no proxy is yet present, the ProxyWatcher should immediately
+	// yield back to its retry loop.
+	select {
+	case <-w.ResetC:
+	case <-time.After(time.Second):
+		t.Fatalf("Timeout waiting for ProxyWatcher reset.")
+	}
+
+	// Add a proxy server.
+	proxy := newProxyServer(t, "proxy1", "127.0.0.1:2023")
+	require.NoError(t, presence.UpsertProxy(proxy))
+
+	// The first event is always the current list of proxies.
+	select {
+	case changeset := <-w.ProxiesC:
+		require.Len(t, changeset, 1)
+		require.Empty(t, resourceDiff(changeset[0], proxy))
+	case <-ctx.Done():
```

Comment:
w.done will be closed when ctx is closed, but ctx might not be closed when w is done.
```suggestion
	case <-w.Done():
```